### PR TITLE
Rewrite aggregate functions with if 

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Rewrite aggregate functions with if expression as argument when logically equivalent. For example, avg(if(cond, col, null)) can be rewritten to avgIf(cond, col). It is helpful in performance. 

``` sql
:) explain syntax select avg(if(number % 2, number, null)) from numbers(100);

EXPLAIN SYNTAX
SELECT avg(if(number % 2, number, NULL))
FROM numbers(100)

Query id: 59d1af55-b195-40db-b279-c31013d72ee2

┌─explain──────────────────────────┐
│ SELECT avgIf(number, number % 2) │
│ FROM numbers(100)                │
└──────────────────────────────────┘

2 rows in set. Elapsed: 0.001 sec. 

```

Speed up by 1.39x after rewriting. 
``` sql
 :) select avgIf(number, rand() % 2) from numbers(100000000); 

SELECT avgIf(number, rand() % 2)
FROM numbers(100000000)

Query id: ff5634d3-6ccb-453c-80b4-18ebaae7afac

┌─avgIf(number, modulo(rand(), 2))─┐
│                50000038.55415332 │
└──────────────────────────────────┘

1 rows in set. Elapsed: 0.849 sec. Processed 100.03 million rows, 800.21 MB (117.88 million rows/s., 943.03 MB/s.)

:) select avg(if(rand() % 2, number, null)) from numbers(100000000);

SELECT avg(if(rand() % 2, number, NULL))
FROM numbers(100000000)

Query id: baacbeb6-2e75-4f66-b014-7bc0c5093c66

┌─avg(if(modulo(rand(), 2), number, NULL))─┐
│                        50000758.86703222 │
└──────────────────────────────────────────┘

1 rows in set. Elapsed: 1.184 sec. Processed 100.03 million rows, 800.21 MB (84.50 million rows/s., 676.02 MB/s.)
```

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by vdimir — CHANGES_REQUESTED** [review]
@taiyang-li thank you for your effort!

Now we are migrating to the new infrastructure for query analysis and trying to avoid introducing AST-based optimizations because it's usually error-prone.
It's better to implement such optimization on top of _query tree_, it will work with `allow_experimental_analyzer`, which is not production-ready yet. But eventually, old infrastructure with AST-based optimizations will be removed, so it's better to go a bit ahead :)
Check out the _QueryTreePassManager_ section at https://github.com/ClickHouse/ClickHouse/issues/42648 for the examples.

**Comment by taiyang-li** [comment]
> @taiyang-li thank you for your effort!
> 
> Now we are migrating to the new infrastructure for query analysis and trying to avoid introducing AST-based optimizations because it's usually error-prone. It's better to implement such optimization on top of _query tree_, it will work with `allow_experimental_analyzer`, which is not production-ready yet. But eventually, old infrastructure with AST-based optimizations will be removed, so it's better to go a bit ahead :) Check out the _QueryTreePassManager_ section at #42648 for the examples.

@vdimir  done.

File: tests/queries/0_stateless/02518_rewrite_aggregate_function_with_if.sql [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+explain syntax select avg(if(number % 2, number, null)) from numbers(100);
+explain syntax select avg(if(number % 2, null, number)) from numbers(100);
+
+explain syntax select quantiles(0.5, 0.9, 0.99)(if(number % 2, number, null)) from numbers(100);
+explain syntax select quantiles(0.5, 0.9, 0.99)(if(number % 2, null, number)) from numbers(100);
+
+set allow_experimental_analyzer = true;
+
+explain syntax select sum(if(number % 2, number, 0)) from numbers(100);
```
Comment (vdimir) [inline tests/queries/0_stateless/02518_rewrite_aggregate_function_with_if.sql:31]: `explain syntax` doesn't show the changes, it's visible in `EXPLAIN QUERY TREE run_passes = 1 SELECT ...`

**Review by vdimir — COMMENTED** [review]
I suggest not introducing AST optimization at all. It's deprecated due to error proneness. 

Code of pass lgtm, but need to fix build and.

@kitaisreal @novikd Do you want to take a look at the `RewriteAggregateFunctionWithIfPass` ?

**Comment by taiyang-li** [comment]
> I suggest not introducing AST optimization at all. It's deprecated due to error proneness.
> 
> Code of pass lgtm, but need to fix build and.
> 
> @kitaisreal @novikd Do you want to take a look at the `RewriteAggregateFunctionWithIfPass` ?

Ast optimizer is already removed.

**Comment by taiyang-li** [comment]
failed tests not related to this pr .

**Review by vdimir — APPROVED** [review]

**Comment by vdimir** [comment]
@taiyang-li, could you please resolve a conflict (better rebase, but merge is also ok if there're too many conflicts)? I can't push to branch or resolve via web interface for unclear reasons.

**Review by novikd — APPROVED** [review]
LGTM, but requires adding documentation before merging.

**Comment by taiyang-li** [comment]
@vdimir @novikd Sorry to bother you, do you think this pr is ok to merge now?

**Pull request merged**